algorithms.go

```diff
@@ -251,13 +265,22 @@ func tokenBucketNewItem(ctx context.Context, s Store, c Cache, r *RateLimitReq)
 		ResetTime: expire,
 	}
 
+	// if we are doing a check during atomic chaining then remaining number is actually the limit
```

Comment:
I would like to see a separation of concerns around checking a bucket vs. updating the bucket.  Because now there's two places where the bucket's counters are updated as a shared responsibility.  Can we have the update logic pulled out and called by gubernator's behavior logic and not by `tokenBucket`/`leakyBucket`?

Reply:
yeah that's a good point, I will move the logic to persist hits into a new function

Reply:
update: I've decoupled the update logic from the actual algorithms into a new function which is invoked from `peer_client.go` following the actual "check" using the algorithm in the case the request is not part of an atomic chain. hopefully this is the kind of separation you were thinking of. I have also extracted the cache and store loading logic into a common function.